### PR TITLE
Added bug report template generator

### PR DESCRIPTION
**Pull request by Kukovec**
- [ ] ~Tests added for any new code~
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] ~Documentation added for any new functionality~
- [ ] ~Entry added to [UNRELEASED.md](./UNRELEASED.md) for any new functionality~

Part 1 of addressing #943:
A bug report file is now automatically generated on `FailureMessage` in the run directory.

TODO: Add documentation/unreleased

### CODE REVIEW

**Comment by shonfeder** [comment]
I believe we can construct a URL that will open an issue on GitHub, with a parameterized template, rather than pointing users towards a file. Do you think this would be a nicer UI?

See: https://docs.github.com/en/enterprise-server@3.1/issues/tracking-your-work-with-issues/creating-an-issue#creating-an-issue-from-a-url-query

File: tla-io/src/main/resources/issueTemplate.md
Comment (shonfeder) [inline tla-io/src/main/resources/issueTemplate.md:12]: Any added value to making this a standalone file that has to be included in the resources that you have in mind?

My initial thought is that this adds some unneeded complexity (an additional resource, then adhoc templating logic), but it seems our usecase could be served just as well with a function that has this string in and just uses Scala's string interpolation for templating. What do you think?

**Review by shonfeder — CHANGES_REQUESTED** [review]
That makes sense, as far as motive for not using the URL to create an issue (tho I'll note that just following the URL doesn't automatically open an issue, the user would still have the chance to remove any sensitive info they wanted to).

I do have one question about the implementation details, regarding how the file location is threaded through, and one suggestion about the approach that leads us to implement an adhoc templating.

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (Kukovec) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:108]: Not really, since this is not an output file, but the input specification file passed by the user. It's used to read the specification string, to copy into the issue body.

File: tla-io/src/main/resources/issueTemplate.md
Comment (Kukovec) [reply tla-io/src/main/resources/issueTemplate.md:12]: You could just have a string, that's true, but the string would be messy to write, and it's easy to edit the `.md` file as a standalone file to get the desired formatting. I guess string interpolation is better than replacement, so the question is really whether we intend to change the bug report template at all. If the answer is "not really", I suppose we can copy the file into a string.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:35]: I don't understand the reason fro dropping the filename from the command. Does it do any harm to include it?

**Review by shonfeder — CHANGES_REQUESTED** [review]

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (Kukovec) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:108]: No, this file is the `example.tla` in e.g. `check example.tla`, which can be anywhere on the filesystem and doesn't get copied into the rundir at any point.

File: tla-io/src/main/resources/issueTemplate.md
Comment (Kukovec) [reply tla-io/src/main/resources/issueTemplate.md:12]: I'm ambivalent on this, if you'd prefer a string it's easy enough to include.

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (shonfeder) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:108]: I understand that. But the only reason you need to pass it in here is in order to use it for generating the bug report output, no? Anyhow, this isn’t a critical thing. Just seems a bit off to me to be passing in additional arguments here when we already have access to the data via the output manager, which we need anyhow in the bug report template generator.

Comment (Kukovec) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:108]: The thing is, the `OutputManager` doesn't know about this file, because it's not otherwise related to outputs.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:67]: `OutputManager.runDir` does not exist.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:31]: I tried making such generic functionality part of the original `OutputManager` PR. This was the comment on it: https://github.com/informalsystems/apalache/pull/1036#discussion_r728367369

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:40]: That's true.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:35]: No harm, but it's redundant information, since it won't help the person reading the issue, as you can't resolve, e.g. `C:\Users\User\Documents\Specs\xyz.tla` on your machine.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:31]: That comment was advocating against a method that was just a wrapper around `Option.map`, but the relevant logic here is about constructing a path relative to the configured run directory and reading in a file, and then closing the file. It seems like this logic could be nicely encapsulated (along with closing the file) in the same way that we have the `withWriter` utilities? As I noted in the linked comment:

> I could understand the utility here if they did more, i.e., tracked state some how, or avoided having to expose the constants in Name, but as is, they read a bit like unhelpful misdirection to me.

The functionality here of constructing paths, reading input, ensuring the file is closed, is exactly the kind of additional logic and book keeping that I think would justify adding a method.

Also, now that I look at the logic of this file, shouldn't it rather return an `Option[String]` if the file cannot be found or read? Defaulting to an `emptyString` if the file doesn't exist seems nonideal, and could lead to weird outputs with things that are expecting to actually get content.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:35]: But we could hit errors due to file names or something right? My rule of thumb is generally something like: don't do extra work to remove information that might later be useful, unless it should be a secret or is causing harm.

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (Kukovec) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:108]: That seems sensible, but at the same time, probably scope creep for this particular PR. You should make an issue for it though, since I can see it being relevant again with the other part of #943, where the error explainer displays the contents of the input file line where the issue is, for which it will need access to the source file too.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:67]: This is from a recent commit merged into unstable that isn't on this branch yet. I guess we can use that after I rebase in unstable.

Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:67]: Right. All our PR's should assume the code available in unstable I think.

Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:67]: Ah, but I understand. This was only just slipped in with the latest changes to the OutputManager I think. Make sense it wasn't used initially!

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:25]: We shouldn't read this from a file at all, actually. It is available in memory the whole time, so let's just get this data from: https://github.com/informalsystems/apalache/blob/2b736c9ffd874d583de0c1e57ecd100bc251b517/mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/General.scala#L43-L44

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (shonfeder) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:107]: I think it should only a take a few lines of code to make the proposed change, but if you'd rather, perhaps leave a note on #1069 that this should be corrected as part of that work, and I can take care of it later this week?

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (shonfeder) [inline mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:363]:
```suggestion
                s"Please report an issue at $ISSUES_LINK: $e\nA bug report has been generated at [$absPath]."
```

Iiuc, we are generating the report using the template?

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:21]: Following up here: Do agree it makes sense to move this into the output manager and correct the return type?

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:25]: Following up here: We don't need to read this value from a file, cause it's already stored in memory in the the CLI parser instance.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:54]: I won't block on this if you feel that there is value in using the `format` syntax, but I feel like this would be quite a bit more readable if it used named variables in the string interpolation, which the function could take as arguments.

**Review by shonfeder — CHANGES_REQUESTED** [review]

File: mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala
Comment (Kukovec) [reply mod-tool/src/main/scala/at/forsyte/apalache/tla/Tool.scala:363]: Technically, the user is still meant to fill in "Expected Behavior". I'll add a prompt to do that in the stdout string.

**Comment by shonfeder** [comment]
Whenever you're ready for me to have another look, please re-request review by clicking the little recycle symbol by my name in the requested reviewers :)

File: tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala:63]: I don't understand why this is added in this PR. It doesn't seem to be used anywhere?

File: tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala
Comment (shonfeder) [inline tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala:61]: Why should it only be called on TLA files? Why not also store the lines from other kinds of input?

**Review by shonfeder — APPROVED** [review]
Two questions, but not all blocking. LGTM!

Nice that we have removed even more IO operations by caching the source file!

File: tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala:61]: Because JSON file lines don't correspond to anything sensible (in fact, the `location` fields in the JSON reference the original spec's line numbers) and tnt isn't implemented yet.

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:17]: You mean as a log/stdout message or as a replacement for `""` in `getOrElse`?
The latter sounds like a good idea, the former not so much. The only scenario where this returns `""` is when you call Apalache on a `.json` file. In that case, there's really no spec to attach to the template (I do not want to encourage people to attach JSON specs in bug reports).

Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala:17]: I meant in the generated bug template, yeah, just so we know what it means if we get an empty string there. Not at all critical. Makes sense about not getting giant json outputs :)

File: tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala
Comment (shonfeder) [reply tla-io/src/main/scala/at/forsyte/apalache/io/OutputManager.scala:61]: The JSON file lines would correspond to the JSON source data, at least. Couldn't we get an error that related to some data coming from a bit of bad json?

> the location fields in the JSON reference the original spec's line numbers

This makes me wonder whether we are asking for trouble by introducing a new way to count line numbers (outside of the one we are currently using for source mapping.....).

Would the correct logic here not then be to parse the spec from the JSON file that was input and save *that* in memory here?

File: tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala
Comment (Kukovec) [reply tla-io/src/main/scala/at/forsyte/apalache/io/ReportGenerator.scala]: OK, it should now print a `<!-- TLA+ specification not found. -->` in the else case.

**Review by shonfeder — APPROVED** [review]
Sorry! I’d meant to approve before too.

**Pull request merged**
